### PR TITLE
Amend/backend super service

### LINKED ISSUE #89: Super-Service Backend
We should create a "base-implementation" for a backend service which we can just extend if a specific service needs additional functionality. Our current services are mostly CRUD though.

### LINKED ISSUE #98: Exchange PUT and POST on all create/update endpoints
PUT and POST should be exchanged on create and update endpoints.
Create endpoints should be POST and update should be PUT.
This needs to be like this because PUT is idempotent however POST isn't.

### LINKED ISSUE #101: Date validation is blocking update
All said, it's a known bug

### PR DESCRIPTION
**Pull request by needToRoll**
This closes #101 , closes #98 , closes #89 

Summary of changes:

- ​Fixed date validation
- Added serverside super-service
- Resolved PUT/POST issue

I made sure to:

- [x] Add and / or update tests
- [x] Run the code formatter (`CTRL + ALT + L`)
- [x] Update any documentation (javadoc, domain model, erd, ...)
- [x] Follow the contribution guidelines

Please review @jmesserli / cc: @outcobra/developers.

### CODE REVIEW

**Comment by needToRoll** [comment]
I have made the requested changes please review it again @jmesserli

File: backend/src/main/kotlin/outcobra/server/controller/TaskController.kt [hunk 1]
```diff
@@ -18,44 +18,39 @@ open class TaskController @Inject constructor(val taskService: TaskService) {
         return taskService.readAllTasks()
     }
 
-    @PutMapping(value = "/task")
-    fun createTask(@RequestBody taskDto: TaskDto): TaskDto {
-        return taskService.createTask(taskDto)
+    @RequestMapping(value = "/task", method = arrayOf(RequestMethod.POST, RequestMethod.PUT))
+    fun saveTask(@RequestBody taskDto: TaskDto): TaskDto {
+        return taskService.save(taskDto)
     }
```
Comment (jmesserli) [inline backend/src/main/kotlin/outcobra/server/controller/TaskController.kt]: `readAllBySemester`

File: backend/src/main/kotlin/outcobra/server/service/base/internal/DefaultBaseService.kt [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+import javax.transaction.Transactional
+
+/**
+ * Default implementation of a basic service to save, read and delete an entity
+ * @see BaseService
+ * @author Florian Bürgi
+ * @since <since>
+ */
+@Transactional
+open class DefaultBaseService<Entity, Dto, out Repo>
```
Comment (jmesserli) [inline backend/src/main/kotlin/outcobra/server/service/base/internal/DefaultBaseService.kt]: `jpaRepository` is not the whole story anymore, maybe just `repository`?

File: backend/src/main/kotlin/outcobra/server/model/mapper/InstitutionMapper.kt [hunk 3]
```diff
@@ -4,13 +4,14 @@ import org.springframework.stereotype.Component
 import outcobra.server.model.Institution
 import outcobra.server.model.dto.InstitutionDto
 import outcobra.server.model.interfaces.Mapper
+import outcobra.server.model.repository.UserRepository
 import outcobra.server.service.UserService
 import javax.inject.Inject
 
 @Component
 open class InstitutionMapper
-@Inject
```
Comment (jmesserli) [inline backend/src/main/kotlin/outcobra/server/model/mapper/InstitutionMapper.kt]: |`line 23`|
|---|
This should still be using the service

File: backend/src/main/kotlin/outcobra/server/service/internal/DefaultInstitutionService.kt [hunk 4]
```diff
@@ -9,36 +9,21 @@ import outcobra.server.model.interfaces.Mapper
...
+import outcobra.server.service.base.internal.DefaultBaseService
 import javax.inject.Inject
 
 @Service
 @Transactional
 open class DefaultInstitutionService
-@Inject constructor(val mapper: Mapper<Institution, InstitutionDto>,
-                    val userService: UserService,
-                    val repository: InstitutionRepository) : InstitutionService {
+@Inject constructor(mapper: Mapper<Institution, InstitutionDto>,
```
Comment (jmesserli) [inline backend/src/main/kotlin/outcobra/server/service/internal/DefaultInstitutionService.kt]: `readAll`

File: backend/src/main/kotlin/outcobra/server/service/internal/DefaultSchoolClassService.kt [hunk 5]
```diff
@@ -17,37 +18,20 @@ import javax.inject.Inject
  */
 @Component
 @Transactional
-open class DefaultSchoolClassService @Inject constructor(val mapper: Mapper<SchoolClass, SchoolClassDto>,
-                                                         val userService: UserService,
```
Comment (jmesserli) [inline backend/src/main/kotlin/outcobra/server/service/internal/DefaultSchoolClassService.kt]: `readAllByUser`

File: backend/src/main/kotlin/outcobra/server/service/internal/DefaultSchoolClassService.kt [hunk 6]
```diff
@@ -17,37 +18,20 @@ import javax.inject.Inject
...
-                                                         val repository: SchoolClassRepository) : SchoolClassService {
+open class DefaultSchoolClassService
+@Inject constructor(mapper: Mapper<SchoolClass, SchoolClassDto>,
+                    repository: SchoolClassRepository,
+                    val userService: UserService)
+    : SchoolClassService, DefaultBaseService<SchoolClass, SchoolClassDto, SchoolClassRepository>(mapper, repository) {
+
```
Comment (jmesserli) [inline backend/src/main/kotlin/outcobra/server/service/internal/DefaultSchoolClassService.kt]: `readAllByInstitution`

File: backend/src/main/kotlin/outcobra/server/service/internal/DefaultSchoolYearService.kt [hunk 7]
```diff
@@ -9,44 +9,29 @@ import outcobra.server.model.dto.SchoolYearDto
...
-                                                        val mapper: Mapper<SchoolYear, SchoolYearDto>,
-                                                        val schoolYearValidator: SchoolYearValidator) : SchoolYearService {
+open class DefaultSchoolYearService
+@Inject constructor(mapper: Mapper<SchoolYear, SchoolYearDto>,
+                    repository: SchoolYearRepository,
+                    val schoolYearValidator: SchoolYearValidator)
+    : SchoolYearService, DefaultBaseService<SchoolYear, SchoolYearDto, SchoolYearRepository>(mapper, repository) {
 
-    override fun createSchoolYear(schoolYearDto: SchoolYearDto): SchoolYearDto {
-        var schoolYear = mapper.fromDto(schoolYearDto)
```
Comment (jmesserli) [inline backend/src/main/kotlin/outcobra/server/service/internal/DefaultSchoolYearService.kt]: `readAllByClass`

File: backend/src/main/kotlin/outcobra/server/service/internal/DefaultSemesterService.kt [hunk 8]
```diff
@@ -10,53 +10,38 @@ import outcobra.server.model.interfaces.Mapper
...
-                                                      val userService: UserService,
-                                                      val mapper: Mapper<Semester, SemesterDto>,
-                                                      val validator: SemesterValidator) : SemesterService {
-
-    override fun createSemester(semesterDto: SemesterDto): SemesterDto {
-        var semester = mapper.fromDto(semesterDto)
+open class DefaultSemesterService
+@Inject constructor(mapper: Mapper<Semester, SemesterDto>,
+                    repository: SemesterRepository,
+                    val userService: UserService,
```
Comment (jmesserli) [inline backend/src/main/kotlin/outcobra/server/service/internal/DefaultSemesterService.kt]: `readAllBySchoolYear`

File: backend/src/main/kotlin/outcobra/server/service/internal/DefaultSubjectService.kt [hunk 9]
```diff
@@ -10,15 +10,19 @@ import outcobra.server.model.repository.SubjectRepository
...
+open class DefaultSubjectService
+@Inject constructor(mapper: Mapper<Subject, SubjectDto>,
+                    repository: SubjectRepository,
+                    val userService: UserService,
+                    val semesterService: SemesterService)
+    : SubjectService, DefaultBaseService<Subject, SubjectDto, SubjectRepository>(mapper, repository) {
+
     override fun readSubjectsByCurrentSemester(): List<SubjectDto> {
         val currentSemesters = semesterService.getCurrentSemester()
         return currentSemesters.flatMap { readAllSubjectsBySemester(it.id) }
```
Comment (jmesserli) [inline backend/src/main/kotlin/outcobra/server/service/internal/DefaultSubjectService.kt]: `readAllBySemester`

File: backend/src/main/kotlin/outcobra/server/service/internal/DefaultSubjectService.kt [hunk 10]
```diff
@@ -27,32 +31,16 @@ open class DefaultSubjectService @Inject constructor(val repository: SubjectRepo
     override fun readAllSubjectsByUser(): List<SubjectDto> {
         val userId = userService.getCurrentUser()?.id
         val filter = QSubject.subject.semester.schoolYear.schoolClass.institution.user.id.eq(userId)
-        return repository.findAll(filter).map { mapper.toDto(it) }
-    }
```
Comment (jmesserli) [inline backend/src/main/kotlin/outcobra/server/service/internal/DefaultSubjectService.kt]: `readAllBySchoolClassId`

File: backend/src/main/kotlin/outcobra/server/service/internal/DefaultSubjectService.kt [hunk 11]
```diff
@@ -10,15 +10,19 @@ import outcobra.server.model.repository.SubjectRepository
...
+import outcobra.server.service.base.internal.DefaultBaseService
 import javax.inject.Inject
 
 
 @Service
 @Transactional
-open class DefaultSubjectService @Inject constructor(val repository: SubjectRepository,
-                                                     val userService: UserService,
-                                                     val semesterService: SemesterService,
-                                                     val mapper: Mapper<Subject, SubjectDto>) : SubjectService {
```
Comment (jmesserli) [inline backend/src/main/kotlin/outcobra/server/service/internal/DefaultSubjectService.kt]: `readByCurrentSemester`

File: backend/src/main/kotlin/outcobra/server/service/internal/DefaultSubjectService.kt [hunk 12]
```diff
@@ -10,15 +10,19 @@ import outcobra.server.model.repository.SubjectRepository
...
+open class DefaultSubjectService
+@Inject constructor(mapper: Mapper<Subject, SubjectDto>,
+                    repository: SubjectRepository,
```
Comment (jmesserli) [inline backend/src/main/kotlin/outcobra/server/service/internal/DefaultSubjectService.kt]: `readAllByUser`

File: backend/src/main/kotlin/outcobra/server/service/internal/DefaultTaskService.kt [hunk 13]
```diff
@@ -13,61 +13,43 @@ import outcobra.server.service.SchoolClassService
...
+    : TaskService, DefaultBaseService<Task, TaskDto, TaskRepository>(mapper, repository) {
 
     override fun readAllTasks(): List<TaskDto> {
         val userId = userService.getCurrentUser()?.id
         val filter = QTask.task.subject.semester.schoolYear.schoolClass.institution.user.id.eq(userId)
-        return repository.findAll(filter).map { mapper.toDto(it) }
+        return jpaRepository.findAll(filter).map { dtoMapper.toDto(it) }
     }
 
-    override fun createTask(taskDto: TaskDto): TaskDto {
```
Comment (jmesserli) [inline backend/src/main/kotlin/outcobra/server/service/internal/DefaultTaskService.kt]: `readAllBySemester`

File: backend/src/main/kotlin/outcobra/server/service/internal/DefaultTaskService.kt [hunk 14]
```diff
@@ -13,61 +13,43 @@ import outcobra.server.service.SchoolClassService
...
 import javax.inject.Inject
 
 @Component
 @Transactional
-open class DefaultTaskService @Inject constructor(val repository: TaskRepository,
-                                                  val mapper: Mapper<Task, TaskDto>,
-                                                  val schoolClassService: SchoolClassService,
-                                                  val subjectService: SubjectService,
-                                                  val userService: UserService) : TaskService {
+open class DefaultTaskService
```
Comment (jmesserli) [inline backend/src/main/kotlin/outcobra/server/service/internal/DefaultTaskService.kt]: `readAll`

**Comment by jmesserli** [comment]
LGTM 👍 thanks @needToRoll.

**Pull request merged**